baremetal-hello.s

```diff
@@ -137,63 +125,100 @@ park:
         j       park
 
 trap_vector:
-        j noop_handler            # offs  0: user software interrupt
-        j noop_handler            # offs  4: supervisor software interrupt
-        j noop_handler            # offs  8: reserved
-        j noop_handler            # offs  c: machine software interrupt
-        j noop_handler            # offs 10: user timer interrupt
-        j noop_handler            # offs 14: supervisor timer interrupt
-        j noop_handler            # offs 18: reserved
-        j timer_handler           # offs 1c: machine timer interrupt
+        j noop_handler                  # offs  0: user software interrupt
+        j noop_handler                  # offs  4: supervisor software interrupt
+        j noop_handler                  # offs  8: reserved
+        j noop_handler                  # offs  c: machine software interrupt
+        j noop_handler                  # offs 10: user timer interrupt
+        j noop_handler                  # offs 14: supervisor timer interrupt
+        j noop_handler                  # offs 18: reserved
+        j timer_handler                 # offs 1c: machine timer interrupt
 
 noop_handler:
         mret
 
 timer_handler:
-# NB: only do timer in RV64 for now, will do RV32 later
-.ifdef RISCV64
         # print a few interesting registers from the timer handler
-        stackalloc_x 6                  # allocate 6 register size arguments on stack
-        csrr    t2, mcause
+        stackalloc_x 7                  # allocate 7 register size arguments on stack
+        csrr    t2, mhartid
         sx      t2, 0,(sp)
-        csrr    t2, mepc
+        csrr    t2, mcause
         sx      t2, 1,(sp)
-        csrr    t2, mtval
+        csrr    t2, mepc
         sx      t2, 2,(sp)
-        csrr    t2, mstatus
+        csrr    t2, mtval
         sx      t2, 3,(sp)
-        li      t2, MTIME
-        ld      t2, 0(t2)
+        csrr    t2, mstatus
         sx      t2, 4,(sp)
-        li      t2, MTIMECMP
-        ld      t2, 0(t2)
+        li      t2, MTIME               # read from memory-mapped mtime register, @TODO: support 64bit in RV32
+        lx      t2, 0,(t2)
         sx      t2, 5,(sp)
+        li      t2, MTIMECMP_BASE
+        csrr    t3, mhartid
+        slli    t3, t3, 3
+        add     t2, t2, t3              # mtimecmp += 8 * mhartid
+        lx      t2, 0,(t2)              # read from memory-mapped mtimecmp_x64bit[mhartid] @TODO: support 64bit in RV32
+        sx      t2, 6,(sp)
         la      a0, print_timer_str
         mv      a1, sp
         call    printf
         stackfree_x 6
```

Comment:
This `stackfree 6` doesn't match the alloc 7 above.

Reply:
fixed